File: 207/Solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        G = {node:[] for node in range(numCourses)}
        for edge in prerequisites:
            G[edge[1]].append(edge[0])
        
        
        def top_sort(G):
            in_degree = {node: 0 for node in G}
            
            for u in G:
                for v in G[u]:
                    in_degree[v] += 1
            
            queue = [node for node in in_degree if in_degree[node] == 0]
            res = []
            while queue:
                cur_node = queue.pop()
                res.append(cur_node)
                
                for v in G[cur_node]:
                    in_degree[v] -= 1
                    if in_degree[v] == 0:
                        queue.append(v)
            
            return True if len(res) == len(G) else False
        
        return top_sort(G)
                
            
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        def generate_graph(prerequisites):
            graph = {}
            for item in prerequisites:
                if item[0] not in graph:
                    graph[item[0]] = []
                
                if item[1] not in graph:
                    graph[item[1]] = []
                
                graph[item[0]].append(item[1])
            
            return graph
        
        graph = generate_graph(prerequisites)
        print(graph)
        
        
        def dfs_helper(graph, root, visited = None):
            if not visited:
                visited = set()
            
            visited.add(root)
            
            for node in graph[root]:
                if node in visited or dfs_helper(graph, node, visited):
                    return True
            
            visited.remove(root)
            return False
                    
        def detect_cycle(graph):
            
            for root in graph:
                if dfs_helper(graph, root):
                    return True
            
            return False
            
        
        return False if detect_cycle(graph) else True


from collections import defaultdict
from collections import deque

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

        if len(prerequisites) == 0:
            return True


        in_degree = {node: 0 for node in range(numCourses)}
        for p in prerequisites:
            in_degree[p[0]] += 1
        zero_degree = deque([k for k, v in in_degree.items() if v == 0])

        visited = set([k for k, v in in_degree.items() if v == 0])

        while zero_degree:
            node = zero_degree.popleft()

            for p in prerequisites:
                if p[1] == node:
                    in_degree[p[0]] -= 1


            for k,v in in_degree.items():
                if v == 0 and k not in visited:
                    zero_degree.append(k)
                    visited.add(k)
        
        
        if sum(in_degree.values()) > 0:
            return False 
        else:
            return True

File: 207/solution.java
class Solution {
    //图
    HashMap<Integer, ArrayList<Integer>> map = new HashMap<>();
    //所有的课
    Boolean flag = false;
    HashSet<Integer> rootSet = new HashSet<>();
    
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if(prerequisites.length == 0) return true;
        if(prerequisites.length == 1 ) return true;
        graph(prerequisites);
        
        if(this.cycle()) return false;
        
        else  return true;
        
                      
    }
    
    public void graph(int[][]courses){
  
        for(int i = 0; i < courses.length; i ++){
            
            if(!map.containsKey(courses[i][0])) map.put(courses[i][0], new ArrayList<>()); 
            if(!map.containsKey(courses[i][1])) map.put(courses[i][1], new ArrayList<>()); 
            
            map.get(courses[i][0]).add(courses[i][1]);
        }

    }
        
    public Boolean DFS( Integer root, HashSet<Integer> visited){
        visited.add(root);
        this.rootSet.add(root);
        for(int child : map.get(root)){
            if(visited.contains(child)) return true;
            if(this.DFS(child,visited)) return true;   
        }
        visited.remove(root);
        return false;
    }

    
   public Boolean cycle(){
       HashSet<Integer> visited = new HashSet<Integer>();
       
       for(int root : map.keySet()){
          if(! this.rootSet.contains(root)){
              if(this.DFS(root,visited))
                return true;
          }
       }
    
        return false;
   }
}